torchtune/modules/early_exit_loss.py

```diff
@@ -0,0 +1,293 @@
+    """
+    if loss_scale_type == LossScaleType.ONE:
+        loss_scales = torch.ones(len(layer_ids), device=layer_ids.device)
+    elif loss_scale_type == LossScaleType.L:
+        loss_scales = torch.Tensor(layer_ids + 1)
+    elif loss_scale_type == LossScaleType.SUM_L:
+        loss_scales = torch.cumsum(layer_ids + 1, dim=0)
+    elif loss_scale_type == LossScaleType.SQRT_L:
+        loss_scales = torch.sqrt(layer_ids + 1)
+    elif loss_scale_type == LossScaleType.INV_L:
+        loss_scales = 1.0 / (layer_ids + 1)
+    elif loss_scale_type == LossScaleType.INV_SQRT_L:
+        loss_scales = torch.reciprocal(torch.sqrt(layer_ids + 1))
+    else:
+        raise ValueError(f"Unsupported loss_scale type {loss_scale_type}")
+
+    loss_scales = loss_scales * torch.where(layer_ids < n_layers - 1, e_scale, 1.0)
+    # normalize loss scales to ensure that their sum is 1.0
+    loss_scales = loss_scales / torch.sum(loss_scales)
+    assert torch.isclose(torch.sum(loss_scales), torch.Tensor([1.0]).to(loss_scales))
+
+    return loss_scales
+
+
+class EarlyExitCurriculumType(str, Enum):
+    NONE = "none"
+    ROTATIONAL = "rot"
+    GRADUAL = "gradual"
+
+
+# TODO: create a base curriculum class that can be used for other aspects, e.g., dropout, datasets, etc.
+class EarlyExitCurriculum:
+    """
+    A curriculum for early exit loss training, which controls which layers to use their hidden states
+    during training.
+    Args:
+        do_output_hidden_states (List[bool]): A list indicating whether each layer's hidden state
+            should be output to calculate their losses.
+        max_steps (int): The maximum number of steps in the curriculum.
+        train_last_layer (bool, optional): Whether to always calculate loss for the last layer. Defaults to True.
+        verbose (bool, optional): Whether to print verbose logs. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        do_output_hidden_states: List[bool],
+        max_steps: int,
+        train_last_layer: bool = True,
+        verbose: bool = False,
+    ):
+        self._init_do_output_hidden_states = do_output_hidden_states
+        self._do_output_hidden_states = do_output_hidden_states
+        self.train_last_layer = train_last_layer
+        self.verbose = verbose
+        self.max_steps = max_steps
+
+    def step(self) -> None:
+        """
+        Perform a step in the curriculum. Should be called at the end of each iteration during training.
+        """
+        pass
+
+    def get(self) -> np.ndarray:
+        """
+        Get the current output hidden states.
+        Returns:
+            np.ndarray: A list indicating whether we should calculate loss for each layer.
+        """
+        do_output_hidden_states = np.copy(self._do_output_hidden_states)
+        # Ensure last layer is trained
+        if self.train_last_layer:
+            do_output_hidden_states[-1] = True
```

Comment:
Out of curiosity, why do we provide special handling for this case? I also wonder whether we could just boil this class down to a protocol similar to [this](https://github.com/pytorch/torchtune/blob/b1aecb19ef5eee949c822e33ef4263807fb6f63e/torchtune/modules/peft/_utils.py#L17), but will defer to you on what you feel is the right way to expose things here.

Reply:
The idea here is that I am trying to ensure that we use the outputs of the last layer of the model. If we don't use the last layer's output, we are essentially pruning layers from the model, and the model's accuracy will drop a lot. So I created this parameter `train_last_layer` and set it to True by default.

There could be one scenario though where we want `train_last_layer` to be False: if we freeze the model's weights, and add early exit heads at different layers. (By head we mean the model's last layer norm and linear layer). However, currently this PR supports shared head only (all early exits use the same head).

I took a look at the protocol in the link provided. What I understood from it is that it is to support adding adapter modules to a model. I think we can use that in the future if we want to support each layer having it's own head, but now we are not adding any heads or layers or adapters to the model. What do you think?

Reply:
Oh yeah the adapter code pointer was more in terms of providing a looser protocol-style structure to inherit from instead of having e.g. the get method and various instance variables defined. It's probably more relevant if we don't need the `train_last_layer` variable since then `get` basically becomes just a getter for `self._do_output_hidden_states`. But if we do need some custom handling that is common to all curricula I think what you're doing here makes sense.

Reply:
I wasn't aware of `Protocol` in Python and I just read some tutorial about it to understand it.
Yes, I see that the curricula classes are in the borderline between classes inheriting from a parent class, and classes implementing a Protocol.

But because this `get()` function does some handling (rather than just returning `self._do_output_hidden_states` it makes a copy of it and sets the last element to `True` if `self.train_last_layer` is `True`) then maybe for now we can keep it as a parent class rather than a protocol?

Reply:
> But because this get() function does some handling (rather than just returning self._do_output_hidden_states it makes a copy of it and sets the last element to True if self.train_last_layer is True) then maybe for now we can keep it as a parent class rather than a protocol?

Yeah agreed, this makes sense to me